### ISSUE TITLE
Waitlist-related capacity checks for group & vol capacity & ACL ratio

### ISSUE THREAD

**Issue by mscholtz** [washingtontrails/vms#448]
In looking at our code, I think we're missing some edge cases when it comes to enforcing capacity.

Currently, for the overall WP capacity, we are enforcing several things.  In addition to enforcing that the total # of registered people doesn't go over capacity, we are:

1. counting people on the waitlist as holding spots, when deciding if new people can register (i.e. if capacity is 10, we have 9 registered, and 2 on the waitlist, new registrants will be waitlisted)

2. doing a separate calculation to determine who can be invited from the waitlist, which considers those who are already invited separately from waitlisted people who haven't yet been invited.  So if capacity is 10, we have 8 registered, and 4 on the waitlist, one of whom has already been invited to take a spot, it will let me invite 1 more person, but not a party of 2 or more (unless they're willing to be split, in which case it will invite 1 of them..)

But the overall capacity isn't the only capacity we're enforcing.  We also have:

* Volunteer Capacity ( separate limit for non-ACLs/CLs)
* Group Capacity for group WPs open to the public
* ACL/Vol ratio, which gets enforced starting at a given date (new vols get turned away until we get more ACLs)

All of the above are only currently being enforced for new registrations, and not taking into account the waitlist.  That is, we're not enforcing 1 and 2 above separately for these other capacities.

Example 1: WP has capacity of 12 but a Volunteer Capacity of 7.  We have 6 vols and 2 leaders registered, and 2 vols on the waitlist.  If a single vol tries to register, ideally we should probably put them on the waitlist, since the 2 vols already on the waitlist got in line first.  Currently this would not happen, the single vol would get the one remaining vol spot.

Example 2: WP has a capacity of 20 but there's a Group with a capacity of 12.  We have 12 group volunteers registered, and 8 non-group people registered, so we're full.   We have 2 group people on the waitlist.  2 non-group people then cancel, so we have 2 spots open, but group is still at capacity.  If we attempt to invite the 2 people from the waitlist, it will let me, because it's not checking against group capacity at that point.  (OTOH, if 2 new group vols try to register it will waitlist them..)

Should these other capacities work the same as the overall capacity when it comes to the waitlist?  If so, we'll need to engineer this a bit better so that the system knows not only how many total people are on the waitlist, but also how many of them are from the group, or are volunteers, etc.

@ebskopin @WaBirder57 can you please prioritize this?  Is this something we need to enforce for volunteer and group capacities?

And does it apply to ACL ratio too?  Does the system need to consider the ratio including the waitlist when deciding if someone can be registered or invited?  (This would be quite complicated, so hoping the answer is no here..)

[note: I'm 99% sure it works this way, but will need to write tests for each of these to make sure if they want to go ahead with this]

**Comment by ebskopin** [comment]
Ok, I'm not sure I have my head totally wrapped around this, but here are my thoughts:

What you describe in Example 1 makes sense to me, and it seems appropriate to offer the single registrant the spot if there is one spot available and 2 on the waitlist. However, this seems to contradict what you described in 1. Either way, I would say that in such a situation, the spot should be available to the first single registrant who comes along.

In Example 2, the folks on the wait-list should be invited in whatever order they arrived, whether group or not.  Once the WP is full, the ratio, while still important, doesn't necessarily need to dictate who comes off the waitlist. 

Hopefully that gives you some idea of prioritization. The answer to your question regarding ACL ratios with regard to waitlists is a definite no.

**Comment by davisagli** [comment]
It seems wrong to me that a new volunteer would automatically end up registered if there are already volunteers on the waitlist. It seems like it'd be best to waitlist the new volunteer in this case so that WTA staff can make the call.

**Comment by mscholtz** [comment]
@davisagli just to be clear: we do sometimes give spots to new people even if there are people on the waitlist.  But only if there's room within the capacity for all of those people.  That's why the system would give them the spot in example 1: within the overall capacity, there's space for both the new registrant and the people on the waitlist.  There's just not room for them all withing the Volunteer Capacity.

But I agree with you, it seems odd to treat this differently.

**Comment by mscholtz** [comment]
@ebskopin I also wanted to pose an Example 3, which is a variant of Example 2, using Volunteer Capacity in place of Group Capacity.  

Example 3: WP has a capacity of 20 but has a Volunteer Capacity of 15. We have 15 vols registered, and 5 CLs/ACLs registered, so we're full. We have 2 vols on the waitlist. 2 ACLs then cancel, so we have 2 spots open, but we're still at our capacity for vols. If we attempt to invite the 2 people from the waitlist, it will let me, because it's not checking against vol capacity at that point. (OTOH, if 2 new vols try to register it will waitlist them..)

Does that concern you any more than example 2 did?

**Comment by ebskopin** [comment]
Yeah, I certainly hear what you are saying regarding not having the one vol sneak under the 2+ on the waitlist. However, I would like to avoid, say, having a group of 4 signup and have that one slot sit empty until the required number of people had dropped  to accommodate the group of 4. If we are working to be fair, then we should probably just waitlist the single volunteer in that instance, and staff can elect to do otherwise. So, as it sounds like was an earlier consensus, we should not allow the system to let smaller groups trump larger ones on the waitlist, and will instead leave that to staff to determine.

To this: "So you're basically saying that other than the total capacity of the program, which we should never exceed, any of the other ratios & capacities should tell vms when to put people on the waitlist, but when it comes to inviting people from the waitlist, at that point it should be up to the user's discretion whether to exceed those." I say yes. The management of the waitlist will be at the users discretion, and we can forge ahead with that in-mind. 

As for example 3: One of the rationales (I believe...) for the Volunteer Capacity vs overall capacity is to establish an acceptable ratio for CLs/ACLs for a given project.  So, in that instance, it might be important that there be a ratio of 3:1 vols/leadership. For that reason, we should hold steady at the 15 volunteers, and they should not be "released" from the waitlist until the ACL spots have been claimed.

**Comment by mscholtz** [comment]
@ebskopin doesn't the 2nd paragraph in your last reply contradict the 3rd paragraph?  Do you mean that in general you want it to work like paragraph 2, but that paragraph 3 should be an exception to that?  If so, I guess the next question should be: are there other exceptions?

Sorry, I know this is really complicated.  If it would help to hop on the phone to discuss, I'm very happy to do that.

**Comment by mscholtz** [comment]
@ebskopin I never heard back from you re my last comment here..

**Comment by ebskopin** [comment]
Yikes, apologies. I would say we stick with what is described in paragraph 2: "other than the total capacity of the program, which we should never exceed, any of the other ratios & capacities should tell vms when to put people on the wait list, but when it comes to inviting people from the wait list, at that point it should be up to the user's discretion whether to exceed those"

Since staff will be actively inviting people from the wait list, they can make the decision to do so or not in instances like you describe in example 3.

**Comment by mscholtz** [comment]
@ebskopin and @WaBirder57 remember this ticket?  I just came across it after it had been buried for a while, because it never got prioritized.  And lately I've been pulling lists of High Priroity tickets, and this of course hasn't been on there.  (I asked you guys to prioritize it when I first posted it, but my fault for not reminding you about it.)

So questions at this point are:

a) how high a priority is this?

a) if it's high priority, can this wait until after launch, or phase 2?  This would be fairly time-consuming, and I'm hoping not to have to do this before launch. Given Elliot's answers above, there's not as much of a problem here as I originally feared.  Of my 3 examples above, only the behavior in example 1 needs to be changed. 

Thanks!
